### ISSUE TITLE
Fix S2743 FP: for static members returning an (external) reference or new instance

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#4851]
### Description
The reasoning is that having static properties/fields on generic classes are a bad thing (when the static/shared) are not related to the generic property. However, if the static property (it can be the case fro a field, but there I see the option for false positives) however, if the `=>` notation is used, and a new instance or external reference is provided, the original concerns are not longer valid; in the first case it can be both a new instance every time, or the shared (singleton) instance is indeed shared, and in the second option does not share anything at all.

I'm in doubt if the `=>` notation should always lead to a compliant solution, but I'm not sure yet.

### Repro steps
``` C#
class Generic<T>
{
    public static string Empty => string.Empty; // FP: Compliant, it only returns an external reference.
    public static object New => new object(); // FP: Compliant, every call returns a new instance.
    public static object Method => SomeMethod(); // Inconclusive 
    
    private static object SomeMethod() => new object();
}
```

**Comment by costin-zaharia-sonarsource** [comment]
Hi @Corniel, I'm not sure I understand your reasoning regarding `=>`. It is just syntactic sugar, and from the functional point of view, it's the same thing. 

Can you elaborate a bit on why you see them differently?

**Comment by Corniel** [comment]
It is syntactic sugar for: `{ get {  return /* something*/  } }`, so every call, is a new return of the external reference. Opposite to `public static Empty = string.Empty`, which defines a new reference.

**Comment by costin-zaharia-sonarsource** [comment]
> It is syntactic sugar for: { get { return /* something*/ } }

My question was why should we treat `=> new object()` differently than `{ get { return new object(); } }`?

**Comment by Corniel** [comment]
> My question was why should we treat `=> new object()` differently than `{ get { return new object(); } }`?
No, but it is functionally different from the assignement, which is what the rule is (should be) about. So yes, also the `{ get }` should be taken into account.

**Comment by costin-zaharia-sonarsource** [comment]
I've added a [PR](https://github.com/SonarSource/sonar-dotnet/pull/4929) to document the false positives. What do you think? Do I miss something?

Overall I do agree with you that issues on properties that are returning a new object from a getter are false positives. If `=>` is used the problem becomes obvious in most cases.

I do think that we should still raise on properties with setter or the ones with just the getter if the initializer is used.

The last case you mentioned is indeed inconclusive. I've found it on external projects too: https://github.com/PiranhaCMS/Piranha/blob/master/Core/Piranha/Data/ActiveRecord.cs#L239-L245.
Usually, when we don't really sure, we don't want to raise to avoid FPs. They can be annoing.

**Comment by Corniel** [comment]
We agree. Indeed, the point is about statics that return an existing reference that is not affected by the generic argument.

**Comment by mary-georgiou-sonarsource** [comment]
This issue seems to have been fixed by this [PR](https://github.com/SonarSource/sonar-dotnet/pull/8364).
